### PR TITLE
Verifiable random function generation and verification functionality

### PR DESCRIPTION
**Pull request by coventry**
Implementation of VRFs, (roughly) following [this draft standard](https://tools.ietf.org/html/draft-goldbe-vrf-01#section-5). See design notes in `VRF.sol`.

---
This change is [Reviewable](https://reviewable.io/reviews/smartcontractkit/chainlink/894)

### CODE REVIEW

**Comment by coventry** [comment]
BTW, the relevant files here are `Curve.sol` and `Curve.js`. I'll clean up the rest of the experiments as things firm up.

File: solidity/contracts/crypto/BigModExp.sol [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  function wordBigModExp(uint256 _base, uint256 _exponent, uint256 _modulus)
+    public view returns (uint256, bool) {
+    uint256 callResult = 0;
+    uint256 callValue;
+    assembly {
+      // Lay out the arguments in memory, as the bigModExp contract expects.
+      // Since each argument is one word long, we increment the address by 0x20
+      // each time. First three arguments are the argument lengths, which are
+      // each 256 bits.
+      mstore(0x40, 0x20)  // Length of base argument
```
Comment (j16r) [inline solidity/contracts/crypto/BigModExp.sol]: Not sure you're supposed to just overwrite 0x40 like this, since it's the free space pointer?

**Comment by coventry** [comment]
Thanks, you're right. That was one of the earlier experiments.

File: solidity/contracts/crypto/BigModExp.sol (hunk 1 above)
Comment (coventry) [reply solidity/contracts/crypto/BigModExp.sol]: Thanks, you're right. That was one of the earlier experiments. It's better to create arrays in solidity code and use those as pointers in the assembly, like I do in `Curve.sol`.

**Comment by coventry** [comment]
This is still very rough, but I got rid of the out of date files, and added what I have of the go, so far.

File: services/vrf/vrf.go [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+func SquareRoot(x *big.Int) *big.Int {
+	return exp(x, sqrtPower, P)
+}
+
+// YSquared returns x^3+3 mod P
+func YSquared(x *big.Int) *big.Int {
+	return mod(add(exp(x, three, P), three), P)
+}
+
+// IsCurveXOrdinate returns true iff there is y s.t. y^2=x^3+3
```
Comment (se3000) [inline services/vrf/vrf.go]: TIL: [iff](https://en.wikipedia.org/wiki/If_and_only_if) 👍

File: services/vrf/vrf.go [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+}
+
+// NewCurvePoint returns the bn256.G1 point corresponding to (x, y)
+func NewCurvePoint(x, y *big.Int) (*bn256.G1, error) {
+	p := new(bn256.G1)
+	packed, err := packUint256s(x, y)
+	if err != nil {
+		return &bn256.G1{}, err
+	}
+	b, err := p.Unmarshal(packed) // Unmarshal's comment lies about return type
```
Comment (se3000) [inline services/vrf/vrf.go]: PR bucket list.

File: services/vrf/vrf.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+	if len(b) != 0 {
+		panic(fmt.Errorf(
```
Comment (se3000) [inline services/vrf/vrf.go]: Stay calm, no need to panic. Preferably we'd just return an error.

If you were relying on it to ensure Generator is properly initialized, we could wrap the call to NewCurvePoint in a private function just for initialization.

File: solidity/contracts/VRF.sol [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+ *   by the specification of RS2ECP) that the x-ordinate should be rejected if
+ *   it is greater than the prime modulus. The modulus for Alt-BN128 is a couple
+ *   of bits shorter than 2**256-1, so hashToCurve masks those bits (see
+ *   zqHash.)
+ *
+ *   The spec also requires the y ordinate of the hashToCurve to be negated if y
+ *   is odd. See http://www.secg.org/sec1-v2.pdf#page=17 . This appears to be
+ *   purely for compositional purposes: They need to specify some way to
+ *   determine y given x, and happened to pick that one. Here, y is determined
+ *   as (x^3+3)^{(p+1)/4, instead.}
```
Comment (se3000) [inline solidity/contracts/VRF.sol]: Interesting. Is this an alternative to the additional v parameter in secp256k1 signatures?

File: solidity/contracts/VRF.sol [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    bigModExpContractInputs[0] = wordLengthBytes;  // Length of _base
+    bigModExpContractInputs[1] = wordLengthBytes;  // Length of _exponent
+    bigModExpContractInputs[2] = wordLengthBytes;  // Length of _modulus
+    bigModExpContractInputs[3] = _base;
+    bigModExpContractInputs[4] = _exponent;
+    bigModExpContractInputs[5] = _modulus;
+    uint256[1] memory output;
+    assembly {
+      callResult :=
+        staticcall(                          // Let bigmodexp use arbitrary gas
```
Comment (se3000) [inline solidity/contracts/VRF.sol]: Is there not a fixed gas amount for this call?

File: solidity/contracts/VRF.sol [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+  // _scalar*p, in the curve group. Reverts if p not on curve.
+  function scalarMul(uint256[2] memory _p, uint256 _scalar)
+    public view returns (uint256[2] memory xyOutput) {
+    uint256 callResult;
+    uint256[3] memory inputs;
+    inputs[0] = _p[0];
+    inputs[1] = _p[1];
+    inputs[2] = _scalar;
+    assembly {
+      callResult := // See comments on bigModExp assembly.
```
Comment (se3000) [inline solidity/contracts/VRF.sol]: This is actually a place where I think more comments would help. A little duplication is worth it, the purpose of the values isn't going to change and it removes the need to jump around the code when reading.

One day we'll get constants in assembly.🤞

File: services/vrf/vrf_test.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
+package vrf
```
Comment (se3000) [inline services/vrf/vrf_test.go]: 🙌 Love that these tests are all in the same package. Would be great to move towards this more often.

File: services/vrf/vrf.go [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+
+import (
+	"bytes"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	// NB: this curve is actually alt-bn128, not bn256!
+	"github.com/ethereum/go-ethereum/crypto/bn256"
+
```
[the remark this answers is not in the document]
Comment (coventry) [reply services/vrf/vrf.go]: > refer to the already defined variables, like cloudflare.P

There's something weird about their build system... `P` etc. don't seem to have made it into the `bn256` package, despite being in `go-ethereum/crypto/bn256/cloudflare/constants.go`. That's the main reason they're duplicated here.

```
grc go test -v services/vrf/vrf*.go                                                                                                                                      
# command-line-arguments 
services/vrf/vrf.go:53:9: undefined: "github.com/smartcontractkit/chainlink/vendor/github.com/ethereum/go-ethereum/crypto/bn256".P 
```

**Comment by coventry** [comment]
@se3000 , thanks for the feedback. I've added some improvements to the golang API. As we discussed, I'll implement the RSA version, next.

**Comment by coventry** [comment]
I'm closing this, in line with the plan we made during IPM. Will open a new one shortly.

**Pull request closed without merging**